module/VuFind/src/VuFind/RecordDriver/AbstractBase.php

```diff
@@ -228,6 +235,53 @@ public function deleteTags($user, $tags)
         }
     }
 
+    /**
+     * Get rating information for this record.
+     *
+     * Returns an array with the following keys:
+     *
+     * rating - average rating (0-100)
+     * count  - count of ratings
+     *
+     * @param ?int $userId User ID, or null for all users
+     *
+     * @return array
+     */
+    public function getRatingData(?int $userId = null)
+    {
+        // Cache data since comments list may ask for same information repeatedly:
+        $cacheKey = $userId ?? '-';
+        if (!isset($this->ratingCache[$cacheKey])) {
+            $table = $this->getDbTable('Ratings');
+            $this->ratingCache[$cacheKey] = $table->getForResource(
+                $this->getUniqueId(),
+                $this->getSourceIdentifier(),
+                $userId
+            );
+        }
+        return $this->ratingCache[$cacheKey];
+    }
+
+    /**
+     * Add or update user's rating for the record.
+     *
+     * @param int $userId ID of the user posting the rating
+     * @param int $rating The user-provided rating
+     *
+     * @return void
+     */
+    public function addOrUpdateRating(int $userId, int $rating): void
+    {
+        // Clear rating cache:
+        $this->ratingCache = [];
```

Comment:
Do you need to clear the whole rating cache here, or could you just update the value for the user directly?

Reply:
It's not very likely that the cache would be hit during the request for update, so updating it here, in any usual case, is not worth it.